CASSANDRA-17048: UUID based sstable generation numbers

jacek-lewandowski opened the pull request.

jeffjirsa commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -101,9 +102,9 @@ private SystemKeyspace()
     public static final String LOCAL = "local";
     public static final String PEERS_V2 = "peers_v2";
-    public static final String PEER_EVENTS_V2 = "peer_events_v2";
+    public static final String PEER_EVENTS_V2 = "peer_events_v2";  // v2 has modified generation column type (v1 - int, v2 - blob), see CASSANDRA-17048
```
Is this comment on the right table?

jacek-lewandowski replied:
oh right, thanks! 🤦🏻

djatnieks commented on src/java/org/apache/cassandra/io/sstable/UUIDBasedSSTableUniqueIdentifier.java:
```diff
@@ -0,0 +1,146 @@
+        /**
+         * Creates a new ULID based identifiers generator.
+         *
+         * @param existingIdentifiers not used by ULID based generator
```
nit: ULID -> UUID

djatnieks commented on src/java/org/apache/cassandra/io/sstable/UUIDBasedSSTableUniqueIdentifier.java:
```diff
@@ -0,0 +1,146 @@
+        @Override
+        public UUIDBasedSSTableUniqueIdentifier fromBytes(@Nonnull ByteBuffer byteBuffer) throws IllegalArgumentException
+        {
+            Preconditions.checkArgument(byteBuffer.remaining() >= 16, "Invalid lenght of the input string - %s < 16", byteBuffer.remaining());
```
nit: typo "length"

smiklosovic commented on conf/cassandra.yaml:
```diff
@@ -1534,3 +1534,10 @@ enable_drop_compact_storage: false
 #		warn_threshold_kb: 0
 #		abort_threshold_kb: 0
+
+# Starting from 4.1 sstables support UUID based generation identifiers. They are disabled by default
```
should not this be in NEWS file in the root too? In section about upgrading. Nobody is going to read the yaml upon upgrade to diff what is new.

jacek-lewandowski replied:
added

djatnieks commented on src/java/org/apache/cassandra/service/CacheService.java:
```diff
@@ -422,7 +426,8 @@ public void serialize(KeyCacheKey key, DataOutputPlus out, ColumnFamilyStore cfs
             ByteArrayUtil.writeWithLength(key.key, out);
-            out.writeInt(key.desc.generation);
+            out.writeInt(Integer.MIN_VALUE); // backwards compatibility for "int based generation only"
+            ByteBufferUtil.writeWithShortLength(key.desc.generation.asBytes(), out);
```
Does this now need to be conditional based on `enable_uuid_generation_identifiers`?

jacek-lewandowski replied:
hmm... yes, perhaps it should be 👍🏻

jacek-lewandowski replied:
actually, we do not need to refer to `DatabaseDescriptor`. We can just write the sequential identifiers the old way.

djatnieks commented on test/unit/org/apache/cassandra/db/SerializationHeaderTest.java:
```diff
@@ -89,7 +90,7 @@ public void testWrittenAsDifferentKind() throws Exception
         {
             BiFunction<TableMetadata, Function<ByteBuffer, Clustering<?>>, Callable<Descriptor>> writer = (schema, clusteringFunction) -> () -> {
-                Descriptor descriptor = new Descriptor(BigFormat.latestVersion, dir, schema.keyspace, schema.name, generation.incrementAndGet(), SSTableFormat.Type.BIG);
+                Descriptor descriptor = new Descriptor(BigFormat.latestVersion, dir, schema.keyspace, schema.name, new SequenceBasedSSTableUniqueIdentifier(generation.incrementAndGet()), SSTableFormat.Type.BIG);
```
nit: maybe use `newSeqGen(generation.incrementAndGet())` for consistency with other tests?

djatnieks commented on src/java/org/apache/cassandra/db/SystemKeyspaceMigrator41.java:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    @VisibleForTesting
+    static void migrateSSTableActivity()
```
Q: Should consideration be given for a possible second upgrade attempt, assuming that `enable_uuid_generation_identifiers` remains `false`? 
e.g. upgrade, then downgrade, then upgrade again
In that case, the _v2 table will exist already and `migrateTable` currently will do nothing - is that ok?

jacek-lewandowski replied:
this is a good point, @belliottsmith mentioned that as well. The suggested solution was to introduce some system value which would denote the last version which the node was running. Fortunately such mechanism already exists.

blerer replied:
Any plan to test that scenario?

djatnieks commented on test/unit/org/apache/cassandra/db/lifecycle/LogTransactionTest.java:
```diff
@@ -1218,7 +1218,7 @@ public void testGetTemporaryFilesThrowsIfCompletingAfterObsoletion() throws Thro
     private static SSTableReader sstable(File dataFolder, ColumnFamilyStore cfs, int generation, int size) throws IOException
     {
-        Descriptor descriptor = new Descriptor(dataFolder, cfs.keyspace.getName(), cfs.getTableName(), generation, SSTableFormat.Type.BIG);
+        Descriptor descriptor = new Descriptor(dataFolder, cfs.keyspace.getName(), cfs.getTableName(), new SequenceBasedSSTableUniqueIdentifier(generation), SSTableFormat.Type.BIG);
```
nit: Another possible `Util.newSeqGen` use here

djatnieks commented on test/unit/org/apache/cassandra/io/sstable/DescriptorTest.java:
```diff
@@ -84,12 +84,12 @@ public void testFromFilenameInLegacyDirectory() throws Exception
     private void testFromFilenameFor(File dir)
     {
-        checkFromFilename(new Descriptor(dir, ksname, cfname, 1, SSTableFormat.Type.BIG));
+        checkFromFilename(new Descriptor(dir, ksname, cfname, new SequenceBasedSSTableUniqueIdentifier(1), SSTableFormat.Type.BIG));
```
nit: and possible `newSeqGen` here as well
I'll stop further suggestions about this :-)

jacek-lewandowski replied:
Actually, that utility is useful when we need to create a generator, but in most cases which are left like that one, we are just creating a fixed id

djatnieks commented on test/unit/org/apache/cassandra/io/sstable/LegacySSTableTest.java:
```diff
@@ -140,14 +141,10 @@ public void tearDown()
-                              SSTableFormat.Type.BIG);
+    protected Descriptor getDescriptor(String legacyVersion, String table) throws IOException
+    {
+        Path file = Files.list(getTableDir(legacyVersion, table).toPath()).findFirst().orElseThrow(() -> new RuntimeException(String.format("No files for verion=%s and table=%s", legacyVersion, table)));
```
nit: spelling `verion` -> `version`

djatnieks commented on test/unit/org/apache/cassandra/io/sstable/SSTableHeaderFixTest.java:
```diff
@@ -790,7 +790,7 @@ private File buildFakeSSTable(File dir, int generation, TableMetadata.Builder co
         {
 
-            Descriptor desc = new Descriptor(version, dir, "ks", "cf", generation, SSTableFormat.Type.BIG);
+            Descriptor desc = new Descriptor(version, dir, "ks", "cf", new SequenceBasedSSTableUniqueIdentifier(generation), SSTableFormat.Type.BIG);
```
Q: Should this test be adapted to use `SSTableUniqueIdentifier` instead only legacy `int` generation values?

djatnieks commented on test/unit/org/apache/cassandra/schema/MockSchema.java:
```diff
@@ -50,6 +57,21 @@
+        return sstableIds.computeIfAbsent(idx, ignored -> sstableIdGenerator.get());
+    }
+
+    public static Collection<Object[]> sstableIdGenerators()
```
Unused?

jacek-lewandowski replied:
not any more

djatnieks commented on test/unit/org/apache/cassandra/schema/MockSchema.java:
```diff
@@ -50,6 +57,21 @@
 
 public class MockSchema
 {
+    public static Supplier<? extends SSTableUniqueIdentifier> sstableIdGenerator = Util.newSeqGen();
```
This could be based on the `SSTableUniqueIdentifierFactory` default or parameterized for both maybe.

jacek-lewandowski replied:
This is just a utility. We cannot parameterize it because it is not a test. The idea is that test classes may use this mechanism.

djatnieks commented on test/unit/org/apache/cassandra/schema/MockSchema.java:
```diff
@@ -177,7 +199,7 @@ public static ColumnFamilyStore newCFS(String ksname, Function<TableMetadata.Bui
     public static ColumnFamilyStore newCFS(TableMetadata metadata)
     {
-        return new ColumnFamilyStore(ks, metadata.name, 0, new TableMetadataRef(metadata), new Directories(metadata), false, false, false);
+        return new ColumnFamilyStore(ks, metadata.name, SequenceBasedSSTableUniqueIdentifier.Builder.instance.generator(Stream.empty()), new TableMetadataRef(metadata), new Directories(metadata), false, false, false);
```
nit: `SequenceBasedSSTableUniqueIdentifier.Builder.instance.generator(Stream.empty())` -> `Util.newSeqGen()`

djatnieks commented on test/unit/org/apache/cassandra/Util.java:
```diff
@@ -850,4 +857,14 @@ public static void setUpgradeFromVersion(String version)
+    public static Supplier<SequenceBasedSSTableUniqueIdentifier> newUlidGen()
+    {
+        return SequenceBasedSSTableUniqueIdentifier.Builder.instance.generator(Stream.empty());
+    }
```
I think this is meant to be:
```
    public static Supplier<UUIDBasedSSTableUniqueIdentifier> newUuidGen()
    {
        return UUIDBasedSSTableUniqueIdentifier.Builder.instance.generator(Stream.empty());
    }
```

i.e.
`newUlidGen` -> `newUuidGen`
`SequenceBasedSSTableUniqueIdentifier` -> `UUIDBasedSSTableUniqueIdentifier`

jacek-lewandowski replied:
yeah, indeed, 🤦🏻

djatnieks reviewed (commented):
Overall looks really good, just a few comments and nits to consider.

djatnieks reviewed (approved):
Looks great, thanks for considering my review suggestions!

blerer commented on NEWS.txt:
```diff
@@ -64,6 +64,10 @@ New features
+    - Starting from 4.1 sstables support UUID based generation identifiers. They are disabled by default in cassandra.yaml
+      because once enabled, there is no easy way to downgrade. When the node is restarted with UUID based generation
+      identifiers enabled, each newly created sstable will have UUID based generation identifier and such files are
+      not readable by previous Cassandra versions. In the future those new identifiers will become enabled by default.
```
Currently we are simply saying that a user can switch to UUID based generation identifiers. I doubt that the benefits of this feature will be obvious for most people. We should brought forward what are the advantages of that change.

blerer commented on src/java/org/apache/cassandra/io/sstable/SSTableUniqueIdentifier.java:
```diff
@@ -0,0 +1,78 @@
+
+/**
+ * Represents a unique identifier in the sstable descriptor filename.
+ * This ensures each sstable files uniqueness in the system.
```
The term system is pretty vague. I think that we should be precise in the level of uniqueness we expect: globaly, cluster-wide ...

blerer commented on src/java/org/apache/cassandra/io/sstable/SSTableUniqueIdentifier.java:
```diff
@@ -0,0 +1,78 @@
+ * This ensures each sstable files uniqueness in the system.
+ * <p>
+ * A new iteration implementation must adhere to the following invariants:
+ * - Must be roughly sortable (for determinism)
```
`roughly` is confusing. The explication should be more precise and should probably contains a justification of why it is required.

blerer commented on CHANGES.txt:
```diff
@@ -1,4 +1,5 @@
 4.1
+ * UUID based sstable generation identifiers (CASSANDRA-17048)
```
Should probably be `Add support for UUID based sstable generation identifiers` we do not enforce the change people will be still free to use the previous approach.

blerer commented on src/java/org/apache/cassandra/io/sstable/SequenceBasedSSTableUniqueIdentifier.java:
```diff
@@ -0,0 +1,122 @@
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
```
nit: the return statement should be on a new line

blerer commented on src/java/org/apache/cassandra/io/sstable/SequenceBasedSSTableUniqueIdentifier.java:
```diff
@@ -0,0 +1,122 @@
+        @Override
+        public SequenceBasedSSTableUniqueIdentifier fromBytes(ByteBuffer bytes) throws IllegalArgumentException
+        {
+            Preconditions.checkArgument(bytes.remaining() >= Integer.BYTES, "Buffer does not have enough data: %s", bytes.remaining());
```
May be we should use `Buffer does not have enough bytes remaining. Expecting: 8 but was : %s`

blerer commented on src/java/org/apache/cassandra/io/sstable/SequenceBasedSSTableUniqueIdentifier.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public int compareTo(SSTableUniqueIdentifier o)
+    {
+        if (o instanceof SequenceBasedSSTableUniqueIdentifier)
+            return Integer.compare(this.generation, ((SequenceBasedSSTableUniqueIdentifier) o).generation);
+
+        return -1;
+    }
```
I am not convinced that comparing different types of `SSTableUniqueIdentifier` within each sub-class is a safe approach.
I would support only comparison with its own type and have a comparator that take care of sorting accross all the sub-types. That would centralise the logic in one place and would make things less error prone if we ever add more sub-types.

jacek-lewandowski replied:
Ok, I'll move it to a separate comparator and remove `Comparable` from the interfaces inherited by `SSTableUniqueIdentifier`

blerer commented on src/java/org/apache/cassandra/io/sstable/SSTableUniqueIdentifierFactory.java:
```diff
@@ -0,0 +1,68 @@
+    public static final SSTableUniqueIdentifierFactory instance = new SSTableUniqueIdentifierFactory();
+
+    /**
+     * Tries to guess the exact identifier type and create an instance of {@link SSTableUniqueIdentifier} using
+     * {@link SSTableUniqueIdentifier.Builder#fromString(String)} from the guessed builder
```
The comment is a bit scary. I would expect the method to return the proper identifier. What if it guess wrongly? 😰

jacek-lewandowski replied:
I'll fix it

blerer commented on src/java/org/apache/cassandra/io/sstable/SSTableUniqueIdentifierFactory.java:
```diff
@@ -0,0 +1,68 @@
+     */
+    public SSTableUniqueIdentifier fromString(String str) throws IllegalArgumentException
+    {
+        SSTableUniqueIdentifier.Builder<?> builder = str.length() == UUIDBasedSSTableUniqueIdentifier.STRING_LEN
```
May be we should add a `isUniqueIdentifier(String)` method to `UUIDBasedSSTableUniqueIdentifier` and make sure that it cannot return `true` for an integer large enough.

jacek-lewandowski replied:
ok

blerer commented on src/java/org/apache/cassandra/io/sstable/SSTableUniqueIdentifierFactory.java:
```diff
@@ -0,0 +1,68 @@
+     */
+    public SSTableUniqueIdentifier fromBytes(ByteBuffer bytes)
+    {
+        SSTableUniqueIdentifier.Builder<?> builder = bytes.remaining() == UUIDBasedSSTableUniqueIdentifier.BYTES_LEN
```
The code seems to expect the buffer to contains only identifier bytes but it is not clearly mentioned. If it is not the case we can end upd with the wrong identifier being returned in a silent way. Same thing if for some reason the buffer contains an unexpected number of bytes (10 for example).

jacek-lewandowski replied:
I'll try to address it somehow

blerer commented on src/java/org/apache/cassandra/io/sstable/UUIDBasedSSTableUniqueIdentifier.java:
```diff
@@ -0,0 +1,146 @@
+    public boolean equals(Object o)
+    {
+
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
```
nit: formatting

blerer commented on test/unit/org/apache/cassandra/io/sstable/SSTableUniqueIdentifierTest.java:
```diff
@@ -0,0 +1,92 @@
+    @Test
+    public void testUUID()
+    {
+        qt().forAll(longs().all(), longs().all()).checkAssert((msb, lsb) -> {
+            msb = (msb & ~0xf000) | 0x1000; // v1
+            UUID uuid = new UUID(msb, lsb);
+            assertThat(uuid.version()).isEqualTo(1);
+            UUIDBasedSSTableUniqueIdentifier id = new UUIDBasedSSTableUniqueIdentifier(uuid);
+            ByteBuffer buf = id.asBytes();
+            UUIDBasedSSTableUniqueIdentifier fromBytes = UUIDBasedSSTableUniqueIdentifier.Builder.instance.fromBytes(buf);
+            assertThat(fromBytes).isEqualTo(id);
+            String s = id.asString();
+            assertThat(s).hasSize(UUIDBasedSSTableUniqueIdentifier.STRING_LEN);
+            assertThat(s).matches(Pattern.compile("[0-9a-z]{4}_[0-9a-z]{4}_[0-9a-z]{18}"));
+            UUIDBasedSSTableUniqueIdentifier fromString = UUIDBasedSSTableUniqueIdentifier.Builder.instance.fromString(s);
+            assertThat(fromString).isEqualTo(id);
+        });
```
The test is a bit hard to read. I would split it in to sub-parts like `testBytesSerialization` and `testStringSerialization`.
It would also be good to check that the underlying UUID is still equals to the original one.

blerer reviewed (commented):
The patch is pretty large it will take me several rounds to go through it. I only scratched some part so far.

blerer commented on test/unit/org/apache/cassandra/cql3/CQLTester.java:
```diff
@@ -603,6 +603,11 @@ public ColumnFamilyStore getCurrentColumnFamilyStore(String keyspace)
              : Keyspace.open(keyspace).getColumnFamilyStore(currentTable);
```
Should be replaced by `getColumnFamilyStore`

jacek-lewandowski replied:
I'm sorry, I don't get - what should be replaced?

adelapena replied:
I think that this should be replaced by a call to the new `CQLTester#getColumnFamilyStore` method right below:
```suggestion
             : getColumnFamilyStore(keyspace, currentTable);
```

blerer reviewed (commented):
I went through the patch and my main concern is testing. I do not feel that we have a proper for coverage  for the new generation numbers or for the case when the user switch from the old generation way to the new way. If this go bad we can endup with some serious data loss.

adelapena commented on src/java/org/apache/cassandra/db/ColumnFamilyStore.java:
```diff
@@ -194,7 +194,7 @@
 
     /* This is used to generate the next index for a SSTable */
-    private final AtomicInteger fileIndexGenerator = new AtomicInteger(0);
+    private final java.util.function.Supplier<? extends SSTableUniqueIdentifier> fileIndexGenerator;
```
Nit: maybe we can extend the import of `com.google.common.base.*` to:
```
import com.google.common.base.Function;
import com.google.common.base.Joiner;
import com.google.common.base.Predicate;
import com.google.common.base.Predicates;
```
And import `java.util.function.Supplier` to use just `Supplier` without the qualified class name.

jacek-lewandowski replied:
oh, interesting, it must a leftover, thanks

adelapena commented on src/java/org/apache/cassandra/db/ColumnFamilyStore.java:
```diff
@@ -373,7 +373,7 @@ public void setCompressionParametersJson(String options)
     public ColumnFamilyStore(Keyspace keyspace,
                              String columnFamilyName,
-                             int generation,
+                             java.util.function.Supplier<? extends SSTableUniqueIdentifier> fileIndexGenerator,
```
This is not related with the changes, but maybe we can remove the unused `offline` argument here.

jacek-lewandowski replied:
I'd not do that in this PR because it may turn out that the lack of usage of `offline` is a bug here

adelapena commented on src/java/org/apache/cassandra/db/ColumnFamilyStore.java:
```diff
@@ -758,19 +746,13 @@ public void loadNewSSTables()
-        do
-        {
-            newDescriptor = new Descriptor(descriptor.version,
+            Descriptor newDescriptor = new Descriptor(descriptor.version,
```
Nit: aligning
```
        Descriptor newDescriptor = new Descriptor(descriptor.version,
                                                  targetDirectory,
                                                  descriptor.ksname,
                                                  descriptor.cfname,
                                                  fileIndexGenerator.get(),
                                                  descriptor.formatType);
```

adelapena commented on src/java/org/apache/cassandra/index/sasi/SASIIndex.java:
```diff
@@ -72,9 +73,7 @@ public SecondaryIndexBuilder getIndexBuildTask(ColumnFamilyStore cfs,
-            NavigableMap<SSTableReader, Map<ColumnMetadata, ColumnIndex>> sstables = new TreeMap<>((a, b) -> {
-                return Integer.compare(a.descriptor.generation, b.descriptor.generation);
-            });
+            NavigableMap<SSTableReader, Map<ColumnMetadata, ColumnIndex>> sstables = new TreeMap<>(Comparator.comparing(t -> t.descriptor.generation, SSTableUniqueIdentifierFactory.COMPARATOR));
```
We can use `SSTableReader.generationComparator`:
```suggestion
            NavigableMap<SSTableReader, Map<ColumnMetadata, ColumnIndex>> sstables = new TreeMap<>(SSTableReader.generationComparator);
```

adelapena commented on src/java/org/apache/cassandra/index/sasi/SASIIndex.java:
```diff
@@ -113,8 +112,7 @@ public SASIIndex(ColumnFamilyStore baseCfs, IndexMetadata config)
 
-        SortedMap<SSTableReader, Map<ColumnMetadata, ColumnIndex>> toRebuild = new TreeMap<>((a, b)
-                                                -> Integer.compare(a.descriptor.generation, b.descriptor.generation));
+        SortedMap<SSTableReader, Map<ColumnMetadata, ColumnIndex>> toRebuild = new TreeMap<>(Comparator.comparing(t -> t.descriptor.generation, SSTableUniqueIdentifierFactory.COMPARATOR));
```
We can use `SSTableReader.generationComparator`:
```suggestion
        SortedMap<SSTableReader, Map<ColumnMetadata, ColumnIndex>> toRebuild = new TreeMap<>(SSTableReader.generationComparator);
```

adelapena commented on test/unit/org/apache/cassandra/cql3/ViewComplexDeletionsTest.java:
```diff
@@ -354,7 +356,7 @@ private void testCommutativeRowDeletion(boolean flush) throws Throwable
             List<String> sstables = cfs.getLiveSSTables()
                                        .stream()
-                                       .sorted((s1, s2) -> s1.descriptor.generation - s2.descriptor.generation)
+                                       .sorted(Comparator.comparing(s -> s.descriptor.generation, SSTableUniqueIdentifierFactory.COMPARATOR))
```
We can use `SSTableReader.generationComparator`:
```suggestion
                                       .sorted(SSTableReader.generationComparator)
```

adelapena commented on src/java/org/apache/cassandra/io/sstable/Descriptor.java:
```diff
@@ -260,14 +263,14 @@ public static Descriptor fromFilename(File file)
+            generation = SSTableUniqueIdentifierFactory.instance.fromString(tokens.get(1));
         }
-        catch (NumberFormatException e)
+        catch (RuntimeException e)
```
Wouldn't it be `IllegalArgumentException`?

jacek-lewandowski replied:
well, there is no way to enforce that IAE will be the only possible exception thrown by that method. I think we should interpret all runtime exceptions the same way and provide the appropriate message when re-throwing

adelapena replied:
Should we do the same with the try-catch with `IllegalArgumentException` for the format part of the name, a few lines below?

adelapena commented on src/java/org/apache/cassandra/io/sstable/SSTableUniqueIdentifier.java:
```diff
@@ -0,0 +1,87 @@
+ * A new implementation must adhere to the following invariants:
+ * - Must be locally sortable - that is, the comparison must reflect the comparison of generation times of identifiers
+ * generated on the same node
+ * - String representation must *not* include the '-' character see {@link Descriptor#filenameSplitter}
```
Nit: `Descriptor#filenameSplitter` is not visible from here. Maybe we can put  the `'-'` separator into a public constant in descriptor, and make `Descriptor#filenameSplitter` and these comments use it?

adelapena commented on src/java/org/apache/cassandra/io/sstable/SSTable.java:
```diff
@@ -101,6 +102,11 @@ protected SSTable(Descriptor descriptor, Set<Component> components, TableMetadat
         this.optimizationStrategy = Objects.requireNonNull(optimizationStrategy);
     }
 
+    public Set<Component> getComponents()
```
Nit: this is only used by tests, we should probably mark it with `@VisibleForTesting`.

adelapena commented on test/unit/org/apache/cassandra/cql3/GcCompactionTest.java:
```diff
@@ -264,7 +267,7 @@ public void testGarbageCollectOrder() throws Throwable
 
         SSTableReader[] tables = cfs.getLiveSSTables().toArray(new SSTableReader[0]);
-        Arrays.sort(tables, (o1, o2) -> Integer.compare(o1.descriptor.generation, o2.descriptor.generation));  // by order of compaction
+        Arrays.sort(tables, Comparator.comparing(o -> o.descriptor.generation, SSTableUniqueIdentifierFactory.COMPARATOR));  // by order of compaction
```
Nit: we can use `SSTableReader.generationComparator`:
```suggestion
        Arrays.sort(tables, SSTableReader.generationComparator);  // by order of compaction
```

adelapena commented on test/unit/org/apache/cassandra/cql3/ViewComplexTest.java:
```diff
@@ -265,8 +267,8 @@ private void testCellTombstoneAndShadowableTombstones(boolean flush) throws Thro
                                        .stream()
-                                       .sorted(Comparator.comparingInt(s -> s.descriptor.generation))
-                                       .map(s -> s.getFilename())
+                                       .sorted(Comparator.comparing(s -> s.descriptor.generation, SSTableUniqueIdentifierFactory.COMPARATOR))
```
Nit: We can use `SSTableReader.generationComparator`:
```suggestion
                                       .sorted(SSTableReader.generationComparator)
```

adelapena commented on test/unit/org/apache/cassandra/db/ColumnFamilyStoreTest.java:
```diff
@@ -266,12 +266,11 @@ public void testBackupAfterFlush() throws Throwable
-            Descriptor desc = new Descriptor(Directories.getBackupsDirectory(existing), KEYSPACE2, CF_STANDARD1, version, SSTableFormat.Type.BIG);
-            for (Component c : new Component[]{ Component.DATA, Component.PRIMARY_INDEX, Component.FILTER, Component.STATS })
+            Descriptor existing = liveSSTable.descriptor;
+            Descriptor desc = new Descriptor(Directories.getBackupsDirectory(existing), KEYSPACE2, CF_STANDARD1, liveSSTable.descriptor.generation, liveSSTable.descriptor.formatType);
```
Nit: can we break this line?
```suggestion
            Descriptor desc = new Descriptor(Directories.getBackupsDirectory(existing), 
                                             KEYSPACE2, 
                                             CF_STANDARD1, 
                                             liveSSTable.descriptor.generation, 
                                             liveSSTable.descriptor.formatType);
```

adelapena commented on test/unit/org/apache/cassandra/db/DirectoriesTest.java:
```diff
@@ -193,13 +222,13 @@ public FakeSnapshot createFakeSnapshot(TableMetadata table, String tag, boolean
     }
 
-    private static List<File> createFakeSSTable(File dir, String cf, int gen) throws IOException
+    private List<File> createFakeSSTable(File dir, String cf, int gen) throws IOException
```
Doesn't throw `IOException`

adelapena commented on test/unit/org/apache/cassandra/db/DirectoriesTest.java:
```diff
@@ -193,13 +222,13 @@ public FakeSnapshot createFakeSnapshot(TableMetadata table, String tag, boolean
     }
 
-    private static List<File> createFakeSSTable(Descriptor desc) throws IOException
+    private List<File> createFakeSSTable(Descriptor desc) throws IOException
```
Doesn't throw `IOException`

adelapena commented on test/unit/org/apache/cassandra/db/DiskBoundaryManagerTest.java:
```diff
@@ -20,8 +20,12 @@
 
 import java.net.UnknownHostException;
 import java.util.List;
+import java.util.stream.Stream;
```
Nit: unused import

adelapena commented on test/unit/org/apache/cassandra/db/DiskBoundaryManagerTest.java:
```diff
@@ -20,8 +20,12 @@
 import com.google.common.collect.Lists;
+
+import org.apache.cassandra.Util;
+import org.apache.cassandra.io.sstable.SequenceBasedSSTableUniqueIdentifier;
```
Nit: unused import

adelapena commented on test/unit/org/apache/cassandra/db/ImportTest.java:
```diff
@@ -27,19 +27,22 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Set;
+import java.util.stream.Stream;
```
Nit: unused import

adelapena commented on test/unit/org/apache/cassandra/db/ImportTest.java:
```diff
@@ -27,19 +27,22 @@
 import org.apache.cassandra.db.lifecycle.LifecycleTransaction;
 import org.apache.cassandra.dht.BootStrapper;
 import org.apache.cassandra.io.sstable.Component;
+import org.apache.cassandra.io.sstable.SequenceBasedSSTableUniqueIdentifier;
```
Nit: unused import

adelapena commented on test/unit/org/apache/cassandra/db/compaction/CompactionStrategyManagerTest.java:
```diff
@@ -28,11 +28,14 @@
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
```
Nit: unused import

adelapena commented on test/unit/org/apache/cassandra/db/compaction/CompactionStrategyManagerTest.java:
```diff
@@ -28,11 +28,14 @@
 import com.google.common.collect.Sets;
 import com.google.common.io.Files;
+
+import org.apache.cassandra.io.sstable.SequenceBasedSSTableUniqueIdentifier;
```
Nit: unused import

adelapena commented on test/unit/org/apache/cassandra/io/sstable/SSTableUniqueIdentifierTest.java:
```diff
@@ -0,0 +1,142 @@
+        assertThat(shuffledIds).isNotEqualTo(ids);
+
+        List<SSTableUniqueIdentifier> sortedIds = new ArrayList<>(shuffledIds);
+        Collections.sort(sortedIds, SSTableUniqueIdentifierFactory.COMPARATOR);
```
Nit: we can use `sortedIds.sort(SSTableUniqueIdentifierFactory.COMPARATOR)`

adelapena commented on test/unit/org/apache/cassandra/io/sstable/SSTableUtils.java:
```diff
@@ -162,7 +163,7 @@ public Context dest(Descriptor dest)
          * Sets the generation number for the generated SSTable. Ignored if "dest()" is set.
          */
-        public Context generation(int generation)
+        public Context generation(SSTableUniqueIdentifier generation)
```
Nit: the JavaDoc for the method should say something like `Sets the identifier for the generated SSTable`

adelapena commented on src/java/org/apache/cassandra/io/sstable/SSTableUniqueIdentifier.java:
```diff
@@ -0,0 +1,87 @@
+ * - must be case-insensitive because the sstables can be stored on case-insensitive file system
+ * <p>
+ */
+public interface SSTableUniqueIdentifier
```
What would you think about naming this `SSTableId` for the sake of brevity, as we do with [`TableId`](https://github.com/apache/cassandra/blob/eb0631646f824a4998556386ebd979ea31874c0f/src/java/org/apache/cassandra/schema/TableId.java)? 

I think that the JavaDoc is clear enough to don't let us miss that this is an unique identifier, and the shorter name seems to help a bit with the long signatures and method calls. I gave it a go in [this commit](https://github.com/adelapena/cassandra/commit/edae13c7a88491e6383fda13ae246392e943749a), if you want to see how it would look like.

jacek-lewandowski replied:
I'm ok with this idea 👍🏻

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/SSTableGenerationTest.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    @Test
+    public void testRestartWithUUIDEnabled() throws IOException, ExecutionException, InterruptedException
```
`ExecutionException` and `InterruptedException` are not thrown. That leaves the import for `ExecutionException` unused.

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/SSTableGenerationTest.java:
```diff
@@ -0,0 +1,177 @@
+                      .withCauseInstanceOf(SystemExitException.class);
+        }
+    }
+
+
```
Super nit: double blank line

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/SSTableGenerationTest.java:
```diff
@@ -0,0 +1,177 @@
+            checkSSTableActivityRow(SSTABLE_ACTIVITY_V2, seqGenId.asBytes(), false);
+            if (expectLegacyTableIsPopulated)
+                checkSSTableActivityRow(LEGACY_SSTABLE_ACTIVITY, seqGenId.generation, false);
+
+
```
Super nit: double blank line

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/SSTableGenerationTest.java:
```diff
@@ -0,0 +1,177 @@
+    private static void checkSSTableActivityRow(String table, Object genId, boolean expectExists)
+    {
+        String cql = "SELECT rate_15m, rate_120m FROM system.%s WHERE keyspace_name=? and columnfamily_name=? and generation=?";
+        UntypedResultSet results = executeInternal(String.format(cql, table), "ks", "tab", genId);
```
Nit: can we assert that `results` is not null?
```suggestion
        UntypedResultSet results = executeInternal(String.format(cql, table), "ks", "tab", genId);
        Assert.assertNotNull(results);
```

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/SSTableGenerationTest.java:
```diff
@@ -0,0 +1,177 @@
+import static org.apache.cassandra.distributed.shared.FutureUtils.waitOn;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class SSTableGenerationTest extends TestBaseImpl
```
Maybe we should add similar JVM upgrade tests testing what happens when upgrading from older versions, wdyt?

jacek-lewandowski replied:
We can, though this test verifies what needs to be verified I think:
- `testRestartWithUUIDEnabled` starts a node with new ids feature disabled, creates some sstables and stops the node, those sstables are verified to conform the old naming format; then the node is restarted, thus this is very similar to upgrading the node where there are existing sstables with old numbering
- `testRestartWithUUIDDisabled` shows that we cannot disable the feature if we have existing sstables generated with UUID

I can add scenario where we just start a node with the feature disabled, create some sstables and restart the node (keep the feature disabled), but this scenario is well covered by many other tests

blerer replied:
What I would like to be sure is that we do not have some untested usecases that can lead to some issue in production. The scenarios that come to my mind are:
* streaming (zerocopy) between 2 nodes. One with the feature enabled the other one with the feature disabled
* compactions running with a mix of old names and new names (for the different compaction strategies)
* restore from a backup (old files only) with only new files or with a mix of new and old
* restore from a backup (old and new files)  with only new files or with a mix of new and old

adelapena replied:
For the streaming scenario I would include a rolling upgrade test with a pre-4.1 node and a 4.1 node with the feature enabled.

jacek-lewandowski replied:
tbh, I added simple streaming tests cases (streaming from a node with mixed sstables to a node with feature disabled and to a node with feature enabled) but adding more does not seem to be relevant - streaming has nothing to do with sstable file names - they are created locally by the stream receiver. Stream header does not transport any information about the identifier and on the receiver side, the sstable writer is always created with a new descriptor

blerer commented:
Regarding the compaction tests and the others within `SSTableIdGenerationTest`. What we have is one partition per SSTable that could hide some of the problems as the rows do not have to be combined with others from other SSTables. We cannot covered everything but having a few update and deletions within different SSTables might lead to a more trustable output.

adelapena commented on src/java/org/apache/cassandra/db/ColumnFamilyStore.java:
```diff
@@ -391,7 +392,7 @@ public ColumnFamilyStore(Keyspace keyspace,
         crcCheckChance = new DefaultValue<>(metadata.get().params.crcCheckChance);
         viewManager = keyspace.viewManager.forTable(metadata.id);
-        fileIndexGenerator.set(generation);
+        this.fileIndexGenerator = fileIndexGenerator;
```
Maybe we can call this `sstableIdGenerator` or `sstableIdProvider`.

adelapena commented on src/java/org/apache/cassandra/io/sstable/Descriptor.java:
```diff
@@ -51,42 +49,44 @@
     public final String ksname;
     public final String cfname;
-    public final int generation;
+    public final SSTableId generation;
```
We could call this attribute `sstableId`, or even `id`, now that `generation` is an attribute of one of the implementations of the more generic `SSTableId` interface. The same applies to the method parameters named `generation`, local variables and mentions in comments.

jacek-lewandowski replied:
hmm.... this will make this commit quite large comparing to what it does; if we really need it, I'd not do that in this PR

adelapena replied:
I don't think it adds that much complexity, and it seems natural to change the names of attributes and parameters in the same commit where we change their type. I gave it a quick try in [this commit](https://github.com/adelapena/cassandra/commit/3afbb0f76f8f0e2d95c8dbe15717f6412d3a2d62), just to see how it would look like.

adelapena replied:
@blerer what do you think about this? Are you ok with renaming `generation` to `id`/`sstableid` in a separate ticket? I guess I can live with it.

adelapena commented on src/java/org/apache/cassandra/io/sstable/Descriptor.java:
```diff
@@ -100,7 +100,7 @@ public Descriptor(Version version, File directory, String ksname, String cfname,
     }
 
-    public Descriptor withGeneration(int newGeneration)
+    public Descriptor withGeneration(SSTableId newGeneration)
```
It seems this method is never called.

adelapena commented on src/java/org/apache/cassandra/io/sstable/Descriptor.java:
```diff
@@ -260,14 +265,14 @@ public static Descriptor fromFilename(File file)
+        catch (RuntimeException e)
         {
-            throw invalidSSTable(name, "the 'generation' part of the name doesn't parse as a number");
+            throw invalidSSTable(name, "the 'generation' part of the name doesn't parse as a valid unique identifier");
```
I think it would be:
```suggestion
            throw invalidSSTable(name, "the 'id' part of the name doesn't parse as a valid unique identifier");
```

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/SSTableIdGenerationTest.java:
```diff
@@ -0,0 +1,229 @@
+        testCompactionStrategiesWithMixedSSTables(SizeTieredCompactionStrategy.class, DateTieredCompactionStrategy.class, TimeWindowCompactionStrategy.class, LeveledCompactionStrategy.class);
+    }
+
+    @SuppressWarnings("unchecked")
+    private void testCompactionStrategiesWithMixedSSTables(Class<? extends AbstractCompactionStrategy>... compactionStrategyClasses) throws Exception
```
Nit: We can mark this method with `final` and use `@SafeVarargs`, so we can get rid of `@SuppressWarnings("unchecked")` here and in the other version of `testCompactionStrategiesWithMixedSSTables`.

jacek-lewandowski replied:
heh, the Java's requirement for a private method to be final is a bit odd

adelapena commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -255,6 +251,18 @@ private SystemKeyspace()
+                "CREATE TABLE %s ("
+                + "keyspace_name text,"
+                + "columnfamily_name text,"
+                + "generation blob,"
```
This column could be named simply `id`

blerer replied:
+1

adelapena replied:
I think this hasn't been addressed yet. By the way, since we are creating a new version of the table, we could rename the `columnfamily_name` column to `table_name`, which is the name that we use in `IndexInfo`, `paxos_repair_history` and `table_estimates`.

adelapena replied:
@jacek-lewandowski I see the new `table_name` column name, which is nice, but we are still missing renaming the `generation` column to `id`.

adelapena commented on src/java/org/apache/cassandra/io/sstable/SSTableId.java:
```diff
@@ -0,0 +1,89 @@
+
+/**
+ * Represents a unique identifier in the sstable descriptor filename.
+ * This ensures each sstable file uniqueness in the entire Cassandra cluster.
```
This might be true with the UUID-based implementation, but I understand that the sequence-based implementation only ensures the uniqueness of the generated `SSTableId`s for the local node, doesn't it?

adelapena commented on test/unit/org/apache/cassandra/io/sstable/SSTableIdTest.java:
```diff
@@ -0,0 +1,141 @@
+        testSSTableIdProperties(UUIDBasedSSTableId.Builder.instance);
+    }
+
+    private void testSSTableIdProperties(SSTableId.Builder<?> builder)
```
I understand that another property of `SSTableId.Builder` is that the ids produced by a generator are unique even with concurrent calls to the supplier returned by `SSTableId.Builder#generator`. Maybe we could test this by concurrently generating some ids with a generator, and then verifying that there aren't any repetitions? For example with something like:
```java
private static <T extends SSTableId> void test(SSTableId.Builder<T> builder) throws Throwable
{
    Set<SSTableId> ids = new CopyOnWriteArraySet<>();
    Supplier<T> generator = builder.generator(ids.stream());
    ExecutorService service = Executors.newFixedThreadPool(NUM_THREADS);
    CountDownLatch latch = new CountDownLatch(1);
    for (int i = 0; i < NUM_THREADS; i++)
    {
        service.submit(() -> {
            try
            {
                Assert.assertTrue(latch.await(1, TimeUnit.MINUTES));
            }
            catch (Throwable t)
            {
                throw new RuntimeException(t);
            }
            for (int j = 0; j < IDS_PER_THREAD; j++)
                ids.add(generator.get());
        });
    }
    latch.countDown();
    service.shutdown();
    Assert.assertTrue(service.awaitTermination(10, TimeUnit.MINUTES));
    Assert.assertEquals(NUM_THREADS * IDS_PER_THREAD, ids.size());
}
```

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/SSTableIdGenerationTest.java:
```diff
@@ -0,0 +1,229 @@
+    @Test
+    public void testCompactionStrategiesWithMixedSSTables() throws Exception
+    {
+        testCompactionStrategiesWithMixedSSTables(SizeTieredCompactionStrategy.class, DateTieredCompactionStrategy.class, TimeWindowCompactionStrategy.class, LeveledCompactionStrategy.class);
```
Nit: can we break this long line?
```suggestion
        testCompactionStrategiesWithMixedSSTables(SizeTieredCompactionStrategy.class,
                                                  DateTieredCompactionStrategy.class,
                                                  TimeWindowCompactionStrategy.class,
                                                  LeveledCompactionStrategy.class);
```

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/SSTableIdGenerationTest.java:
```diff
@@ -0,0 +1,229 @@
+            {
+                String tableName = "tbl_" + compactionStrategyClass.getSimpleName().toLowerCase();
+
+                cluster.schemaChange(String.format("CREATE TABLE %s.%s (pk int, v int, primary key (pk)) WITH compaction = {'class':'%s', 'enabled': 'false'}", KEYSPACE, tableName, compactionStrategyClass.getCanonicalName()));
```
Nit: can we break this long line? Maybe something like:
```suggestion
                cluster.schemaChange(String.format("CREATE TABLE %s.%s (pk int, v int, primary key (pk)) " +
                                                   "WITH compaction = {'class':'%s', 'enabled': 'false'}",
                                                   KEYSPACE, tableName, compactionStrategyClass.getCanonicalName()));
```

jacek-lewandowski commented:
@adelapena @blerer can we move forward with this?

adelapena commented:
@jacek-lewandowski it seems that the patch is still breaking multiple tests (see [CircleCI results](https://app.circleci.com/pipelines/github/adelapena/cassandra/1226/workflows/0ccfbbcd-08b0-4454-ba32-f3dc63b58f73)).

adelapena commented on src/java/org/apache/cassandra/io/sstable/AbstractSSTableSimpleWriter.java:
```diff
@@ -82,34 +82,22 @@ protected SSTableTxnWriter createWriter()
-
-            return false;
-        });
-        int maxGen = generation.getAndIncrement();
```
I think `AbstractSSTableSimpleWriter#generation` is not used anymore.

jacek-lewandowski replied:
why? It is still used

adelapena replied:
My bad, I should have confused it with a local var or something, I don't remember but it looks good.

jacek-lewandowski commented:
@adelapena it has many merge conflicts, I was waiting for your approval before force pushing after rebase because that would break your reviews; I was going to look into tests after that

jacek-lewandowski committed:
Implement sstable generation identifier as uuid

Patch by Jacek Lewandowski, reviewed by Dan Jatniecks, Benjamin Lerer, Andrés de la Peña for CASSANDRA-17048

adelapena commented on NEWS.txt:
```diff
@@ -78,6 +78,12 @@ New features
+      the node to create sstables without any prior knowledge about the existing sstables in the data directory.
+      The feature is disabled by default in cassandra.yaml because once enabled, there is no easy way to downgrade.
+      When the node is restarted with UUID based generation identifiers enabled, each newly created sstable will have
+      UUID based generation identifier and such files are not readable by previous Cassandra versions. In the future
```
```suggestion
      a UUID based generation identifier and such files are not readable by previous Cassandra versions. In the future
```

adelapena commented on conf/cassandra.yaml:
```diff
@@ -932,6 +932,13 @@ compaction_throughput_mb_per_sec: 64
 
+# Starting from 4.1 sstables support UUID based generation identifiers. They are disabled by default
+# because once enabled, there is no easy way to downgrade. When the node is restarted with this option
+# set to true, each newly created sstable will have UUID based generation identifier and such files are
```
```suggestion
# set to true, each newly created sstable will have a UUID based generation identifier and such files are
```

adelapena commented on src/java/org/apache/cassandra/db/Directories.java:
```diff
@@ -1202,6 +1202,24 @@ public static boolean isSecondaryIndexFolder(File dir)
 
+    /**
+     * Initializes the sstable unique identifier generator using a provided builder for this instance of directories.
+     * If the uid builder needs that, sstables in these directories are listed to provide the existing identifiers to
```
```suggestion
     * If the id builder needs that, sstables in these directories are listed to provide the existing identifiers to
```

adelapena commented on conf/cassandra.yaml:
```diff
@@ -932,6 +932,13 @@ compaction_throughput_mb_per_sec: 64
+# set to true, each newly created sstable will have UUID based generation identifier and such files are
+# not readable by previous Cassandra versions. At some point, this option will become true by default
+# and eventually get removed from the configuration.
+enable_uuid_generation_identifiers: false
```
I'm wondering whether we should place this properties in the section for experimental features. @blerer wdyt?

blerer replied:
It is tricky because we cannot really say to people: "use that functionality it is experimental and by the way there is no going back".

blerer reviewed (approved):
The patch looks good. There is only a couple of nits left.

adelapena commented on src/java/org/apache/cassandra/io/sstable/SSTableIdFactory.java:
```diff
@@ -0,0 +1,97 @@
+    @SuppressWarnings("unchecked")
+    public SSTableId.Builder<SSTableId> defaultBuilder()
+    {
+        SSTableId.Builder<? extends SSTableId> builder = DatabaseDescriptor.isUUIDGenerationIdentifiersEnabled()
+                                                                                     ? UUIDBasedSSTableId.Builder.instance
+                                                                                     : SequenceBasedSSTableId.Builder.instance;
```
Nit: alignement
```suggestion
        SSTableId.Builder<? extends SSTableId> builder = DatabaseDescriptor.isUUIDGenerationIdentifiersEnabled()
                                                         ? UUIDBasedSSTableId.Builder.instance
                                                         : SequenceBasedSSTableId.Builder.instance;
```

adelapena commented on src/java/org/apache/cassandra/service/StartupChecks.java:
```diff
@@ -569,6 +586,14 @@ public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) th
+                                           "UUID generation identifiers are disabled but some sstables have been " +
+                                           "created with UUID generation identifier. You have to either delete those " +
+                                           "sstables (if you are ok with loosing that data on this node, and then let " +
+                                           "the repair bring them back) or enabled UUID based generation identifers in " +
```
We are not sure that repair is going to be able to bring that data back, right?

jacek-lewandowski replied:
I'll replace it with:

```java
                throw new StartupException(StartupException.ERR_WRONG_CONFIG,
                                           "UUID generation identifiers are disabled but some sstables have been " +
                                           "created with UUID generation identifier. You have to either delete those " +
                                           "sstables or enabled UUID based generation identifers in cassandra.yaml " +
                                           "(enable_uuid_generation_identifiers). The list of affected sstables is: " +
                                           Joiner.on(", ").join(withIllegalGenId) + ". If you decide to delete sstables, " +
                                           "and have that data replicated over other healthy nodes, those will be brought" +
                                           "back during repair");
```

adelapena commented on test/distributed/org/apache/cassandra/distributed/shared/ClusterUtils.java:
```diff
@@ -858,4 +860,23 @@ public String toString()
+    {
+        System.setSecurityManager(new SecurityManager()
+        {
+            public void checkExit(int status)
```
Nit: I think the three overridden methods need `@Override` tags

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/SSTableIdGenerationTest.java:
```diff
@@ -0,0 +1,515 @@
+    public static void beforeClass() throws Throwable
+    {
+        TestBaseImpl.beforeClass();
+        ClusterUtils.preventSystemExit();
```
I'd add a comment about why we need this

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/SSTableIdGenerationTest.java:
```diff
@@ -0,0 +1,515 @@
+
+    /**
+     * The purpose of this test is to verify that we can compact using the given strategy the mix of sstables created
+     * with sequential id and with uuid. Then we verify whether the number of rows.
```
Nit: `Then we verify whether the number of rows.` seems like an unfinished sentence, maybe it's just `Then we verify the number of rows.`?

adelapena commented on test/unit/org/apache/cassandra/cql3/ViewComplexDeletionsTest.java:
```diff
@@ -19,7 +19,7 @@
 
 import java.util.Arrays;
-import java.util.Comparator;
+import java.util.Collection;
```
Nit: unused import

adelapena commented on test/unit/org/apache/cassandra/cql3/ViewComplexTest.java:
```diff
@@ -21,6 +21,7 @@
 import java.nio.ByteBuffer;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collection;
```
Nit: this new import and the other three below are unneeded.

adelapena commented on test/unit/org/apache/cassandra/cql3/ViewComplexTombstoneTest.java:
```diff
@@ -28,6 +28,7 @@
 import org.apache.cassandra.db.ColumnFamilyStore;
 import org.apache.cassandra.db.Keyspace;
 import org.apache.cassandra.db.compaction.CompactionManager;
+import org.apache.cassandra.io.sstable.*;
```
```suggestion
import org.apache.cassandra.io.sstable.SSTableIdFactory;;
```

jacek-lewandowski committed:
apply review comments

adelapena reviewed (approved):
The PR looks good to me. The only part that I don't see that clear is how we mix the terms generation and sstable id, which IMO makes things a bit more confusing than they were before. I'm ok with leaving this for a followup ticket if everyone agrees to it.

I don't see CI results for the last changes, so I've started a run here:
* [j8](https://app.circleci.com/pipelines/github/adelapena/cassandra/1386/workflows/27af9796-120c-4cca-b714-8c2b09fb713f)
* [j11](https://app.circleci.com/pipelines/github/adelapena/cassandra/1386/workflows/c2bcd60a-d040-4500-a7f1-5cf6a5634752)

jacek-lewandowski committed:
rename generation -> id in sstable activity table

jacek-lewandowski committed:
rename generation -> id in the code

jacek-lewandowski committed:
DO NOT MERGE - CircleCI config

jacek-lewandowski commented:
https://app.circleci.com/pipelines/github/jacek-lewandowski/cassandra/207/workflows/b6247ece-53de-4949-aa3f-32849a1204a7
https://app.circleci.com/pipelines/github/jacek-lewandowski/cassandra/207/workflows/181f0856-3b92-4f7d-9646-c9e840ebbc9e
